sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala

```diff
@@ -1092,14 +1092,23 @@ private[hive] object HiveClientImpl extends Logging {
       hiveTable.setViewExpandedText(t)
     }
 
+    // hive may convert schema into lower cases while bucketSpec will not
+    // only convert if case not match
+    def restoreHiveBucketSpecColNames(schema: StructType, names: Seq[String]): Seq[String] = {
+      names.map { name =>
+        schema.find(col => SQLConf.get.resolver(col.name, name)).map(_.name).getOrElse(name)
+      }
+    }
+
     table.bucketSpec match {
       case Some(bucketSpec) if !HiveExternalCatalog.isDatasourceTable(table) =>
         hiveTable.setNumBuckets(bucketSpec.numBuckets)
-        hiveTable.setBucketCols(bucketSpec.bucketColumnNames.toList.asJava)
+        hiveTable.setBucketCols(
+          restoreHiveBucketSpecColNames(table.schema, bucketSpec.bucketColumnNames).toList.asJava)
 
         if (bucketSpec.sortColumnNames.nonEmpty) {
           hiveTable.setSortCols(
-            bucketSpec.sortColumnNames
+            restoreHiveBucketSpecColNames(table.schema, bucketSpec.sortColumnNames)
```

Comment:
Sorry I still can't understand how the bug happens.

In this `toHiveTable` method, the input `CatalogTable` should guarantee sanity: the partition/bucket column names should match the schema.

Reply:
org/apache/spark/sql/hive/client/HiveClient.scala
final def getPartitions(
      db: String,
      table: String,
      partialSpec: Option[TablePartitionSpec]): Seq[CatalogTablePartition] = {
    getPartitions(getTable(db, table), partialSpec)
  }

in this method, we get table from metasotre and pass it into getPartitions
and then call toHiveTable to convert a catalog table into HiveTable.

Reply:
So there is a necessary hive table -> `CatalogTable` -> hive table convertion?

Reply:
this convertion is not unnecessary since the hive client interface require a CatalogTable.

Reply:
Can we change the hive client interface? For example
```
trait HiveClient {

  def getRawTableOption(dbName: String, tableName: String): Option[HiveTable]
  final def getTableOption(dbName: String, tableName: String): Option[CatalogTable] = {
    getRawTableOption(dbName, tableName).map(convertHiveTableToCatalogTable)
  }
  ...
  def getPartitionOption(
      table: HiveTable,
      spec: TablePartitionSpec): Option[CatalogTablePartition] = {
    getPartitions(getRawTable(db, table), partialSpec)
  }
  ...
}
```

Reply:
this is not the only place that have this issue. if we set spark.sql.statistics.size.autoUpdate.enabled=true, you can see this issue as well. for alter table, we have to do a catalogTable->HiveTable conversion

Reply:
catalogTable->HiveTable is fine, as long as the catalogTable is correctly initialized. The problem I see here is, we get catalogTable by `HiveClient.getTable` which doesn't go through the intialization logic in `HiveExternalCatalog`

Reply:
@AngersZhuuuu can you take this over?

Reply:
> @AngersZhuuuu can you take this over?

Sure.